File: Algorithms & Data Structures/Sorting/SelectionSort.java
import static Helper.ListUtils.*;
import static Helper.UnsortedListGenerator.*;

public class SelectionSort {

    public static void main(String[] args) {
        int[] list = generateUnsortedList(30, 1000);

        printHeaders("Before Selection Sort");
        printList(list);

        printHeaders("After Selection Sort");
        printList(performSelectionSort(list));
    }


    /* 
        Selection Sort

        Worst Case runtime: O(n^2)
        Best Case runtime:  O(n)
    */
    private static int[] performSelectionSort(int[] list) {
        int smallestNumberIndex;

        for (int i = 0; i < list.length; i++){
            smallestNumberIndex = i;

            for (int j = i + 1; j < list.length; j++) 
                if (list[j] < list[smallestNumberIndex]) 
                    smallestNumberIndex = j;
            swap(list, i, smallestNumberIndex);
        } 
                    
        return list;        
    }
}

File: Algorithms & Data Structures/Sorting - JavaScript/SelectionSort.js
// ========= Imports =============
const swap          = require("./Helper/ListUtils").swap
const isSorted      = require("./Helper/ListUtils").isSorted
const printHeaders  = require("./Helper/ListUtils").printHeaders
const printList     = require("./Helper/ListUtils").printList

const generateFixedSizeUnsortedList = require("./Helper/UnsortedListGenerator").generateFixedSizeUnsortedList

/* 
Selection Sort

Time Complexity
---------------------
Worst   : O(n^2)

Average : O(n^2)

Best    : O(n^2)
*/
performSelectionSort = (list) => {
    let smallestNumberIndex;
    
    for (let i = 0; i < list.length; i++){
        smallestNumberIndex = i;
        
        for (let j = i + 1; j < list.length; j++) 
            if (list[j] < list[smallestNumberIndex]) 
                smallestNumberIndex = j;
        swap(list, i, smallestNumberIndex);
    } 
    
    return list;        
}


main = () => {
    let list = generateFixedSizeUnsortedList();

    printHeaders("Before Selection Sort");
    printList(list);

    printHeaders("After Selection Sort");
    printList(performSelectionSort(list));

    if (isSorted(list))
        printHeaders("HOORAY!");
    else
        printHeaders("Fail :(");
}

main()
